pkg/osvscanner/customgitignore/walk_up_to_root.go

```diff
@@ -0,0 +1,234 @@
+package customgitignore
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-billy/v5/osfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/format/gitignore"
+)
+
+// Takes a dir and processes .gitignore files from it.
+//
+// This uses go-git under the hood and returns a slice
+// of go-git's gitignore.Pattern structs.
+//
+// The actual parsing is inteded to be similar to how tools
+// rg work, but means that `path` may not necessarily be
+// the root of a git repo, and can produces these parsing
+// behaviours:
+//
+// - parsing file:
+// 	- find the file's dir and then run the following rules...
+//
+// - parsing dir at the root of a git-repo
+// 	- with recursive flag:
+// 		- read all .gitignore files in repo
+// 		- read .git/info/exclude for repo
+// 	- without recursive flag:
+// 		- read .gitignore files in repo-root
+// 		- read .git/info/exclude for repo
+//
+// - parsing dir within, a git-repo
+// 	- with recursive flag:
+// 		- read .gitignore in start dir
+// 		- read all .gitignore files in child-dirs
+// 		- read all .gitignore files in parent dirs up to and including repo-root
+// 			(ie only dirs that are an ancestor)
+// 		- read .git/info/exclude for repo
+//  - NOTE: the dir you're passing in directly could be a dir that is ignored
+//      by a parent's .gitignore or the per-repo exclude file; in this
+//      case, we still process the dir's .gitignore file and, possibly, but
+//      not it's sub-dirs
+//
+// 	- without recursive flag:
+// 		- read .gitignore in start dir
+// 		- read all .gitignore files in parent dirs up to and including repo-root
+// 			(ie only dirs that are an ancestor)
+// 		- read .git/info/exclude for repo
+//
+// - parsing a plain dir: .gitignore files are ignored
+//
+// In all cases any dirs matched by a previously read
+// .gitignore are skipped, unless it's the path (ie directly
+// supplied by the user).
+func ParseGitIgnores(path string, recursive bool) (returnPs []gitignore.Pattern, repoRootPath string, err error) {
```

Comment:
pretty sure that lint is disabled, and we're using these in a few places already - overall I think they can have their uses, but that you should explicitly list them in the `return` calls to make it clear (which @robramsaynz is doing), and not use them that often

Reply:
The thing is that you're not just labeling the returns but actually creating variables, which are used when you don't have an explicit return value and the linter will not report that - that's why Go itself recommends their use only for short functions, because it's easy to get them lost (especially as part of refactors) and they can harm readability within the function itself

There is actually a linter that outright bans them which we've disabled because we're agreed they are fine in small doses, but I wouldn't use them extensively and would even go so far as to say the general default convention in Go is to not use them (though happy to be disagreed with on that).

On top of that a lot of the time (including here) you're not actually using the named variable so you're throwing away a small allocation everytime these functions are used.

I don't think you should be using it on this large function, and looking over your smaller functions I don't think any of them really get much benefit either as they can't take advantage of the default variable value

Reply:
done

Reply:
I don't really have a strong preference on this one, though I think for the purposes of documenting what the return values are we can just put it in the function comment/docstring, avoiding the extra allocation (though I wonder if the go compiler is smart enough to just not do the allocation by default).